Develop no6

takeuchiyui opened the pull request:
No6のファイルを追加しました

tasobue committed:
Create Sheet1

tasobue committed:
Create Thisworkbook

tasobue committed:
Delete Sheet1

tasobue committed:
Delete Thisworkbook

tasobue committed:
Create Sheet1

tasobue committed:
Create module2

tasobue committed:
Create Thisworkbook

takeuchiyui committed:
DateDiff関数追加

DateDiff関数を追加しました

takeuchiyui committed:
変更なし

変更なし

takeuchiyui committed:
変更なし

変更なし

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -0,0 +1,103 @@
+
+
+'DateDiff関数を使う
+'DateDiff（"yyyymmdd",変数,変数+4）??
```
`DateDiff(interval,date1,date2[,firstdayofweek][,firstweekofyear])`

引数intervalには、date1とdate2の間隔を計算するための単位を指定します。
引数date1とdate2には、間隔を計算する２つの日付を指定します。
引数firstdayofweekは省略可能です。週の始まりの曜日を指定します。
引数firstweekofyearは省略可能です。年度の第１週を指定します。

今回は
`DateDiff("d",開始日,終了日)`
でいいです。

ちなみにDateDiffする目的は？

takeuchiyui replied:
DateDiff関数は開始日から終了日までを取得してくれるので、自分のやりたいことと一致していると思いました。
私の考えでは、「今日の日付を取得している変数」と新しく取得する「今日の日付+4日を取得している変数」を使い
DateDiffに当てはめれば選択した期間の日付を取得できると思いました。先ほどまでは、Datediffに直接指定する考え方でしたが変数を使って指定したらいいと思いました

tasobue replied:
```
Sub Sample()
    Dim Date1 As Date, Date2 As Date
    Date1 = #4/1/97#        '97年4月1日
    Date2 = #6/27/97#       '97年6月27日
    MsgBox DateDiff("m", Date1, Date2)      '２ヶ月を返します
    MsgBox DateDiff("d", Date1, Date2)      '87日を返します
    MsgBox DateDiff("ww", Date1, Date2)     '12週を返します
End Sub
```

DateDiffは引数に渡した日付の差を日、月、年などの単位でかえしてくれるだけす。

tasobue commented:
先にコメントを使ってどういうコードを書いていくか全体図を見せてもらっていいですか？
どういう考え方をしているか知ったほうが効率がいいので。

takeuchiyui committed:
テキスト2から日付取得処理を追加

テキスト2から日付取得処理を追加しました

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -12,28 +15,44 @@ Private Sub TextBox1_LostFocus()
+    newday = Format(targetDate, "yyyymmdd")
+    
+    'テキスト2の日付取得
+    nowdate2 = DateAdd("yyyymmdd", 4, Date)
```
これだと当日日付に4日足してるだけで,テキスト2の日付は使われてないですよ。

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -12,28 +15,44 @@ Private Sub TextBox1_LostFocus()
     
-    '// 年月日（スラッシュ区切り）
-    strDate = Format(targetDate, "yyyymmdd")
+    newday = adddate
```
adddateは宣言されてないからエラーになるんじゃない？

tasobue commented:
やりかたとしては、こんな感じがいいんじゃない？
前提：開始日が1/1　終了日が1/4

DateDiffで差分をとり4（または3？）という数値を取得する。

差分の長さを持った配列を用意する

差分の回数だけ取得すべき日付を配列にセットしていく
　配列には[1/1, 1/2, 1/3, 1/4]がセットされる

配列をoutput()に渡す

outputは受け取った配列の回数だけファイルの読み込み、エクセルへの出力を行う
　outputは引数として配列を受け取れるように変える必要あり。
　

takeuchiyui committed:
配列格納処理を追加

配列を使ったやり方を追加しました

tasobue commented:
コメントに対する返答をお願いします！！

tasobue commented:
あと、実際どこまでできているかを教えてもらっていいですか？
ソースだけ見てもどこで困ってるか正確にわかりません。

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -23,21 +26,25 @@ Private Sub TextBox1_LostFocus()
+    
+    '開始と終了を上記変数を指定して、差分をとり配列に格納
+    '宣言の型が合っていないせいかエラーが表示される
+    '差分をとってくれるのでdate型だと思う
```
https://docs.microsoft.com/ja-jp/office/vba/language/reference/user-interface-help/datediff-function

2 つの指定した日付の時間間隔を指定する **Variant (Long) を返します。**

きちんと調べましょう。なんとなくで標準のライブラリを使ってもエラーになりますよ！

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -23,21 +26,25 @@ Private Sub TextBox1_LostFocus()
-    'テキスト2の日付取得
-    nowdate2 = DateAdd("yyyymmdd", 4, Date)
+    'テキスト１から＋4日分を取得
+    nowdate2 = DateAdd("d", 4, newday)
```
4日足す意味は？

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -23,21 +26,25 @@ Private Sub TextBox1_LostFocus()
+  
     'テキスト1の日付取得
-    newday = Format(targetDate, "yyyymmdd")
+    newday = Format(targetDate, "yyyy/mm/dd")
```
Formatする必要あるの？

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -23,21 +26,25 @@ Private Sub TextBox1_LostFocus()
+    '差分をとってくれるのでdate型だと思う
+    preparation = DateDiff("ｄ", newday, nowdate2)
+    '区切りが必要なのでsplitを使う
+     strDate = Split(preparation, vbCrLf)
```
preparationに入るのは数値（Long)なので、Splitすることができません。
配列を用意したいのであれば以下のようになります
```
`Dim 配列名(添字の上限値) As データ型名

 Dim targetDates(separate) As String '※separateはintに変換しないとエラーになるかも
`
```

tasobue commented:
上記の繰り返しになりますがコミットする際に必ずソースだけでなく進捗状況も教えてください。

takeuchiyui commented:
全体の処理の手順
1.テキスト１の値を取得　変数に格納
2.テキスト２の値を取得　変数に格納
3.テキスト１とテキスト２の差分をとる　Datediff関数
4.３の差分を配列に入れる
5.モジュール２の型をvaliant型にする？

tasobue commented:
>4.３の差分を配列に入れる
これをもっと細かく

tasobue commented:
>5.モジュール２の型をvaliant型にする？
モジュール2は配列を受け取ることになるので配列型になります。VariantでOKですか？

takeuchiyui commented:
４.DateDiff関数した結果を配列に入れて、配列を再定義する（ReDim）

tasobue commented:
DateDiff関数した結果は差分の数値だけです。

takeuchiyui commented:
４．差分の結果を変数に格納して、数分だけFor文でまわして日付を取得する

tasobue commented:
```
for(i = 0; i < 差分の数値-2; i ++){
   dim new_date = DateAdd("d", 開始日, i)
   結果を入れる配列[i] = new_date
}
```

takeuchiyui committed:
処理の１～３の処理追加

テキストの値を格納した変数を使い、DateDiff関数を行った→ここまでは完了
DateDiffの第一引数でエラーが出てします。yyyymmddは×　ｄもみ〇（自分の取得したい日付にならない）
その次のFor文を使い差分分だけデータを回して値を取得することができていない

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -25,22 +26,24 @@ Private Sub TextBox1_LostFocus()
+    preparation = DateDiff("d", targetDate, targetDate2)
+    
+    i = 1
+    
```
For文の前にstrDateの配列の長さを日付の差分の価で再定義してください

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -25,22 +26,24 @@ Private Sub TextBox1_LostFocus()
+    
+    For i = i To preparation
+    '差分を取得して配列に入れる
+        strDate(i) = DateDiff("yyyymmdd", targetDate, targetDate2)
```
strDate に入れるのはDateAddを使ってください。

tasobue replied:
DateDiffの返す値を具体的に理解できていないようです。
記事を読み込むか、まったく別のVBAファイルを用意してどのように動くかを確かめるということをしてください。

takeuchiyui commented:
・Redim strDate(preparation)をして再定義する
・DateAddして+1ずつ日付を取得したい
    （targetDateに格納している値+1）

tasobue commented:
いいと思います

tasobue commented:
output側が配列を受け取った後の処理も考えてみましょう

takeuchiyui commented:
output側の配列受け取りの際、valiant型でなければstrDate()が型になる？

tasobue commented:
そうですね。strDate()でいいんじゃないかな？

tasobue commented:
output側も配列を受け取ったら配列の長さだけループを回してファイルの読み込みとセルへの出力を行うことになります。

takeuchiyui committed:
配列に値を格納

・Redim strDate(preparation)をして再定義する→完了
・DateAddして+1ずつ日付を取得したい→未完了
（targetDateに格納している値+1）

再定義した配列の中に格納すると思っていたが、型が違う

takeuchiyui committed:
配列型に変更

配列型に変更しました

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -33,19 +32,21 @@ Private Sub TextBox1_LostFocus()
+          
+    'strDate(preparation)？？格納できない
+    strDate(i) = DateAdd("d", 1, targetDate)
+    i = i + 1
```
Nextステートメントがiをインクリメントしてくれるので不要です

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -33,19 +32,21 @@ Private Sub TextBox1_LostFocus()
-        
+          
+    'strDate(preparation)？？格納できない
+    strDate(i) = DateAdd("d", 1, targetDate)
```
1をiに変えてください

takeuchiyui commented:
１配列に値を格納できたので、モジュール２へ引数を渡したい。
・outputを呼び出す
・output側の型を合わせる　配列なのでByRefを使う
　strDate()はByRefの後に書く？
　
２渡したらループで回してファイル名を作成する

takeuchiyui commented:
２訂正
　モジュールに書いていた今までの処理すべてをループにする

tasobue commented:
配列はByRef, ByValを使う
参照渡し(ByRef)と値渡し(ByVal)の違いで調べてください

outputの引数定義はこういうかきかた
output(ByVal targetdate() String)

tasobue commented:
考え方はあっています。
注意として、ループのたびにセルの出力位置がA1に戻らないように工夫する必要があります。

takeuchiyui committed:
output呼び出し処理追加

output呼び出し処理追加しました。
第一引数に日付
第二引数に要素

上手くわたっているのか分からない

takeuchiyui committed:
引数を受け取る処理

引数の受け取りが上手くいっていない
Public Sub output(ByVal targetDate() As Variant, preparation As Integer)だとエラー（赤字）になる

FilePath = csvFolder & targetDate & "_kintai.csv"
の処理でcsvFolder を定義をしてくださいと表示される（定義済みなのに　Const csvFolder As String = "C:\work\kintai\")

tasobue commented on VBA/No6/module2:
```diff
@@ -1,4 +1,6 @@
-Public Sub output(targetDate As strDate)
+
+'targetDate()だとエラー　（）がないとエラー出ない。配列でも必要ない？
+Public Sub output(ByVal targetDate As Variant, preparation As Integer)
```
Sheet1側で渡す変数が配列になっていないからです。

tasobue commented on VBA/No6/Sheet1:
```diff
@@ -39,16 +39,15 @@ Private Sub TextBox1_LostFocus()
     
     'output()を呼び出して出力する
-    'Call output(strDate)
+    Call output(strDate(preparation), preparation)
```
配列の渡し方はCall output(**strDate**, preparation)です。

strDate(preparation)だとpreparationがインデックス代わりになって、配列から取得した値をoutputに渡してしまいます。

tasobue commented on VBA/No6/module2:
```diff
@@ -13,6 +15,12 @@ Public Sub output(targetDate As strDate)
     
+    '取得した日付分だけループして出力 第１引数：日付　第２引数：要素数
+    '渡ってきた要素分だけ回す
+    For i = i To preparation
```
ループの回数は配列の長さと同じだからpreparationを使う必要はありませんよね？

takeuchiyui commented:
配列分出力する
１　ループ処理　配列の長さ分だけ行う
２　配列に入っている日付を文字列に変換する処理
３　今のままだとA1にかかれてしまうので一番下の行を取得する
　　Cells(Rows.Count, 1).End(xlUp).Row　が使えそう

tasobue commented:
なんとなく伝わるけどどこの処理の話かを明確に書くようにしてください。
outputの処理全体の流れを日本語にしてみてください。

tasobue commented:
A列の最終行を取得したいなら

`Cells(Rows.Count, 1).End(xlDown).Row　`

じゃない？

takeuchiyui commented:
A列をクリアする
↓
For文でtargetDate（最大値）分回す
↓
文字列変換処理を記入
↓
ファイルパスを作成してFilePathに代入
↓
ファイルパスが存在し無い場合
「ファイルパスが存在しません」と表示して処理を終了
↓
ファイルパスのファイルを開き中身を取得する
↓
一番下の行を取得
↓
For　ファイルの中身の分だけ出力して色を付ける

tasobue commented:
いいですね！やってみましょう。

tasobue commented:
結果記入よろしく

takeuchiyui committed:
日付型から文字列型へ変換

まだは配列の中身を文字列変換できていません。

tasobue commented on VBA/No6/module2:
```diff
@@ -11,93 +9,104 @@ Public Sub output(ByVal targetDate As Variant, preparation As Integer)
-            Cells(d, "A").Interior.Color = RGB(0, 0, 255)
+        '定数を定義してくださいと表示がでる
+        'ファイルパスを作る
+        FilePath = csvFolder & targetDate(i) & "_kintai.csv"
```
targetDate(i)はDate型なので
CStr(targetDate(i))にしましょう。

tasobue commented on VBA/No6/module2:
```diff
@@ -11,93 +9,104 @@ Public Sub output(ByVal targetDate As Variant, preparation As Integer)
-        Close #2
+     '---------------------------------------------------------------
+    
+        d = 1
```
今のままだと、ループのたびに1行目からテキスト内容を入力してしまいます。
dの初期値は、さっき書いていたようにA列の最終行にするのがよいでしょう。

tasobue commented:
大筋は間違ってないので、細かいところ、変数に代入する値やループで値がリセットされてしまうところ、配列のインデックスが適切かなどを注視してください。

takeuchiyui commented:
For i = 0 To UBound(targetDate)   
        fo = CStr(targetDate(i))
        FilePath() = csvFolder & fo & "_kintai.csv"
FilePathを配列にすることですべてのパスを取得できるはずです

takeuchiyui commented:
自分の考えではFor文の範囲が----の間だと考えていますが合っているでしょうか。これで実行しましたが、Nextの位置が間違っているとエラーが表示されました

A列をクリアする
↓
------------------------------------------
For文でtargetDate（最大値）分回す
↓
文字列変換処理を記入
↓
ファイルパスを作成してFilePathに代入
↓
ファイルパスが存在し無い場合
「ファイルパスが存在しません」と表示して処理を終了
↓
ファイルパスのファイルを開き中身を取得する
↓
一番下の行を取得
↓
For　ファイルの中身の分だけ出力して色を付ける
----------------------------------------------------------

tasobue commented:
ループの中でファイルパスを作成しているから、ファイルパスは配列にする必要なくないですか？

takeuchiyui commented:
自分の考えではFor文の範囲が☆の間だと考えていますが合っているでしょうか。これで実行しましたが、Nextの位置が間違っているとエラーが表示されました

A列をクリアする
↓
☆For文でtargetDate（最大値）分回す
↓
文字列変換処理を記入
↓
ファイルパスを作成してFilePathに代入
↓
ファイルパスが存在し無い場合
「ファイルパスが存在しません」と表示して処理を終了
↓
ファイルパスのファイルを開き中身を取得する
↓
一番下の行を取得
↓
☆For　ファイルの中身の分だけ出力して色を付ける

tasobue commented on VBA/No6/module2:
```diff
@@ -11,93 +9,104 @@ Public Sub output(ByVal targetDate As Variant, preparation As Integer)
+            
+            Close #2
+            
+            Next
```
たぶんこのNextが余分。

takeuchiyui committed:
日付分のデータを出力

指定した日付のデータを出力することができました！

tasobue commented:
処理結果のキャプチャを張ってください

tasobue commented:
キャプチャはGitHubにも貼れるよ。
![image](https://user-images.githubusercontent.com/21050636/105286301-59697580-5bf9-11eb-90ca-52f86d26fea4.png)

takeuchiyui commented:
![image](https://user-images.githubusercontent.com/77647999/105286352-6e460900-5bf9-11eb-804b-3142aec3f962.png)

takeuchiyui commented:
はれました。

tasobue commented:
No6は範囲指定がテーマなのでシートのレイアウトを変えましょうか。
「前日」「翌日」は削除。
テキストボックスの左に開始日、終了日と記載してわかりやすくする。
「抽出」ボタンを追加して、押下時にデータ取得を行う。

tasobue commented:
次のゴールはこれです
![image](https://user-images.githubusercontent.com/21050636/105286804-58851380-5bfa-11eb-8fb8-444531d586e2.png)

takeuchiyui commented:
今はやり方を考える時間ですが、もうレイアウト修正をしていっていいでしょうか

tasobue commented:
Google Chatに送ったレイアウトで出力するための手順を考えてください。
レイアウト変更に合わせて出力処理が変わってきます。そのロジックを考えましょう。

あと、GitHub上には会社に関する情報は載せないようにお願いします。インターネット上のだれからでも見れてしまうので。

takeuchiyui commented:
注意します。会社情報とは先ほど送ったExcelのキャプチャのことでしょうか？

takeuchiyui commented:
＞No6は範囲指定がテーマなのでシートのレイアウトを変えましょうか。
「前日」「翌日」は削除。
テキストボックスの左に開始日、終了日と記載してわかりやすくする。
「抽出」ボタンを追加して、押下時にデータ取得を行う。

抽出ボタンなどは付けなくとよいという認識で合っているでしょうか。
チャットで送信していただいたレイアウトにしていくということですね

takeuchiyui commented:
１レイアウト変更　日付、枠線を追加
２csvデータから時間だけを取り出す
３２の値を色を付ける処理の場所に、サインインならばEへ出力と処理を変える

tasobue commented:
Chatに送ったキャプチャのほうに合わせてください。

会社情報はプロジェクトで顧客から受領した資料とか、顧客を推測可能な情報です。

tasobue commented:
基本的にソースコード以外の資料は載せないほうがいいですね。
なので、さっき自分でアップした資料はすぐに消しました。

takeuchiyui commented:
> Chatに送ったキャプチャのほうに合わせてください。
　かしこまりました

> 会社情報はプロジェクトで顧客から受領した資料とか、顧客を推測可能な情報です。
　なぜ処理結果を張り付けてくださいと言ったのですか
　ぼかして張り付けて欲しいという意味だったのでしょうか

takeuchiyui commented:
１レイアウト変更　日付、枠線を追加
２数日から取得を1か月にする（まだ考え中）
３csvデータから時間だけを取り出す
４２の値を色を付ける処理の場所に、サインインならばEへ出力と処理を変える

tasobue commented:
>なぜ処理結果を張り付けてくださいと言ったのですか
処理結果をGitHubに張り付けてとお願いしたのは私の間違いでした。すみません。

tasobue commented:
No6のお題をクリアしたのでmainに反映する。
